File: server-go/internal/handlers/backup.go
package handlers

import (
	"net/http"

	"comeback.ai/server-go/internal/auth"
	"comeback.ai/server-go/internal/db"
	"comeback.ai/server-go/internal/middleware"
	"comeback.ai/server-go/internal/models"

	"go.mongodb.org/mongo-driver/bson"
	"go.mongodb.org/mongo-driver/mongo/options"
)

func requireOwner(next http.HandlerFunc) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		ctx := r.Context()
		if !isOwner(ctx, auth.GetUserID(r)) {
			middleware.Error(w, http.StatusForbidden, "Owner access required")
			return
		}
		next(w, r)
	}
}

// RequireOwnerWrapper adapts requireOwner for route registration.
func RequireOwnerWrapper(next http.HandlerFunc) http.HandlerFunc {
	return requireOwner(next)
}

func BackupStatus(w http.ResponseWriter, r *http.Request) {
	ctx := r.Context()
	users, _ := db.Collection("users").CountDocuments(ctx, bson.M{})
	groups, _ := db.Collection("groups").CountDocuments(ctx, bson.M{})
	tasks, _ := db.Collection("tasks").CountDocuments(ctx, bson.M{})
	uptime := timeNow().Unix()
	middleware.Success(w, map[string]interface{}{
		"users": users, "groups": groups, "tasks": tasks, "uptime": uptime,
		"healthy": true,
	})
}

func BackupRun(w http.ResponseWriter, r *http.Request) {
	// No external email/backup transport configured in this Go port; report result.
	middleware.Success(w, map[string]interface{}{"success": false, "message": "Automated backup transport not configured in Go backend"})
}

func BackupLogsSend(w http.ResponseWriter, r *http.Request) {
	middleware.Success(w, map[string]interface{}{"success": false, "message": "Log transport not configured"})
}

func BackupLogs(w http.ResponseWriter, r *http.Request) {
	limit := queryInt(r, "limit", 100)
	middleware.Success(w, map[string]interface{}{"logs": []string{}, "limit": limit})
}

func BackupNotify(w http.ResponseWriter, r *http.Request) {
	var body struct {
		Subject string `json:"subject"`
		Message string `json:"message"`
	}
	_ = decodeBody(r, &body)
	if body.Subject == "" || body.Message == "" {
		middleware.Error(w, http.StatusBadRequest, "Subject and message required")
		return
	}
	middleware.Success(w, map[string]string{"message": "Notification queued (no transport configured)"})
}

func BackupUsers(w http.ResponseWriter, r *http.Request) {
	ctx := r.Context()
	opts := options.Find().SetSort(bson.D{{Key: "createdAt", Value: -1}}).
		SetProjection(bson.M{"password": 0, "refreshToken": 0})
	cur, err := db.Collection("users").Find(ctx, bson.M{}, opts)
	if err != nil {
		middleware.Error(w, http.StatusInternalServerError, "Failed to load users")
		return
	}
	var users []models.User
	_ = cur.All(ctx, &users)
	out := make([]map[string]interface{}, 0, len(users))
	for _, u := range users {
		out = append(out, map[string]interface{}{
			"id": u.ID.Hex(), "email": u.Email, "username": u.Username, "displayName": u.DisplayName,
			"xp": u.Xp, "level": u.Level, "streak": u.Streak, "provider": u.Provider,
			"isOnline": u.IsOnline, "createdAt": u.CreatedAt,
		})
	}
	middleware.Success(w, map[string]interface{}{"data": out, "total": len(out)})
}

File: server/src/routes/backup.js
import { Router } from 'express';
import { authenticate } from '../middleware/auth.js';
import { catchAsync, AppError } from '../middleware/errorHandler.js';
import User from '../models/User.js';
import { runBackup, sendLogs, getLogs, getSystemStatsForResponse, logActivity } from '../services/backupService.js';
import { notifyOwner } from '../services/emailService.js';

const router = Router();

router.use(authenticate);

const requireOwner = async (req, res, next) => {
  const user = await User.findById(req.user.id);
  if (!user || user.email !== process.env.OWNER_EMAIL) {
    return res.status(403).json({ success: false, message: 'Owner access required' });
  }
  next();
};

router.get('/status', requireOwner, catchAsync(async (req, res) => {
  const stats = await getSystemStatsForResponse();
  res.json({ success: true, data: stats });
}));

router.post('/run', requireOwner, catchAsync(async (req, res) => {
  logActivity('Manual backup triggered');
  const result = await runBackup();
  res.json({ success: result, message: result ? 'Backup sent' : 'Backup failed' });
}));

router.post('/logs', requireOwner, catchAsync(async (req, res) => {
  const result = await sendLogs();
  res.json({ success: result, message: result ? 'Logs sent' : 'Failed to send logs' });
}));

router.get('/logs', requireOwner, catchAsync(async (req, res) => {
  const limit = parseInt(req.query.limit) || 100;
  res.json({ success: true, data: getLogs(limit) });
}));

router.post('/notify', requireOwner, catchAsync(async (req, res) => {
  const { subject, message } = req.body;
  if (!subject || !message) throw new AppError('Subject and message required', 400);
  await notifyOwner(`[Manual] ${subject}`, `<p>${message}</p>`);
  res.json({ success: true, message: 'Notification sent' });
}));

router.get('/users', requireOwner, catchAsync(async (req, res) => {
  const users = await User.find().select('email username displayName xp level streak createdAt provider isOnline lastSeen').sort({ createdAt: -1 }).lean();
  res.json({ success: true, data: users, total: users.length });
}));

export default router;
